### PR TITLE
rsz: replace BfsIterator pull-API usage with local DFS traversal

### PR DESCRIPTION
**Pull request by dsengupta0628**
## Summary

  OpenSTA plans to remove BfsIterator::hasNext()/next(), so this PR moves the four remaining rsz users off that API onto a file-local iterative DFS helper in Resizer.cc:

  - findFanins
  - findFaninRoots
  - findFanouts
  - findClkInverters

  The helper (dfsSearch) reproduces Bfs{Fwd,Bkwd}Iterator::enqueueAdjacentVertices adjacency semantics exactly, including the mode-less SearchPred calls (OR across modes). Explicit stack + visited set: no recursion, terminates on cycles, and drops rsz's
  dependence on BfsIndex::other and levelized queue state. 

  Result sets are unchanged (three functions return unordered sets). findClkInverters now sorts its result by graph level before returning — cloneClkInverter requires upstream-first clone order, previously guaranteed implicitly by BFS level order, now
  enforced explicitly.

 Also adds rmp/gcd_restructure_delay regression test: delay-mode restructure was the sole caller of findFanins and had zero coverage (existing tests only exercise area mode).

## Type of Change
- New feature

## Impact
NA

## Verification
- [x] I have verified that the local build succeeds (`./etc/Build.sh`).
- [x] I have run the relevant tests and they pass.
- [x] My code follows the repository's formatting guidelines.
- [x] I have included tests to prevent regressions.
- [x] **I have signed my commits (DCO).**

### CODE REVIEW

**Comment by dsengupta0628** [comment]
Gemini observations are technically accurate, but I profiled the priorities differently, so I'm declining most of this:

  std::function overhead: The indirect call is real but not measurable here. The per-vertex cost of this traversal is dominated by the visited set insert (std::set, O(log n) with a heap allocation per node) and the SearchPred calls, which are genuinely  virtual and additionally iterate over all modes in their mode-less variants. On top of that, both call paths are cold: restructure is an opt-in command and repair_clock_inverters runs once per CTS. std::function visitors are also the established pattern in rsz (see RepairTargetCollector.hh, Rebuffer.cc), so I'd rather stay consistent than template for an unmeasurable win.

  dir branch in the expansion loop: It's loop-invariant, so it's perfectly predicted after the first iteration - if constexpr would eliminate a branch that's already effectively free.

  Anonymous-namespace free function "to comply with repository guidelines": I couldn't find such a guideline; the existing file-local helpers in Resizer.cc (SearchPredCombLogic, ClkArrivalSearchPred) don't use one either. 

Where I agree: if we want a performance pass on this helper, the actual lever is the visited container - sta::VertexSet is an ordered std::set, and since it's used purely for membership (never iterated), switching to std::unordered_set<Vertex*> is
  determinism-safe and removes the dominant per-edge cost. Templating the seeds parameter to accept any range would also drop the three VertexSet-->vector copies at the call sites. I'd take those two as a follow-up because they're orthogonal to the BFS to DFS  migration this PR is scoped to.

**Comment by dsengupta0628** [comment]
[resizer_dfs_mig.pdf](https://github.com/user-attachments/files/29983003/resizer_dfs_mig.pdf)
Spec for this migration

**Comment by dsengupta0628** [comment]
ORFS https://jenkins.openroad.tools/job/OpenROAD-flow-scripts-Private/job/secure-rsz_funcs_dfs/
secure ci http://secure-ci:8080/view/My/job/SB/job/secure-rsz_funcs_dfs/

**Comment by jhkim-pii** [comment]
@codex review

File: src/rmp/test/gcd_restructure_delay.tcl [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+set_wire_rc -layer metal3
+estimate_parasitics -placement
+
+report_worst_slack
+
+set tiehi "LOGIC1_X1/Z"
+set tielo "LOGIC0_X1/Z"
+
+ord::set_thread_count "3"
+restructure -liberty_file Nangate45/Nangate45_typ.lib -target delay \
```
[the remark this answers is not in the document]
Comment (dsengupta0628) [reply src/rmp/test/gcd_restructure_delay.tcl]: actually I now get a seg fault with -target timing in this testcase! Fixing that now before I update this regression

**Comment by dsengupta0628** [comment]
Added a new test gcd_restructure_delay.tcl to cover findFanins, which the DFS work rewrote.

findFanins sole caller = rmp::Restructure::getBlob in timing mode (-target timing). Zero ORFS exposure, no existing test.

Existing rmp restructure test gcd_restructure.tcl uses -target area --> hits findFaninRoots/findFanouts (via findFaninFanouts), not findFanins. So before this, findFanins had no test at all. 
When I added a test, it exposed a pre-exsting bug

 **Why Restructure.cpp needs to be changed**
  vertexWorstSlackPath reads back worst-slack paths that only exist after an arrival/required search; restructure's searchPreamble computes delays only. Timing-mode restructure was never exercised by a test --> nobody hit the missing search + null path.

So we need a force arrival search like this in non area mode:
 ```
 if (!is_area_mode_) {
    open_sta_->worstSlack(sta::MinMax::max());   // force arrival search
  }
```
  ...
  sta::Path* path = open_sta_->vertexWorstSlackPath(end_point, ...);
After above line, we also need:
 ` if (path == nullptr) { continue; }             // skip unconstrained endpoints`

So,
  - worstSlack(max) forces the arrival/required search to run. Without it, every vertexWorstSlackPath returns null (no search data exists) --> total failure.
  - Even after the search, an individual endpoint can have no worst-slack path: unconstrained endpoints (no required time / no clock-relative constraint). the test I added had exactly this - note [WARNING STA-0441] set_input_delay relative to a clock defined on the  same port/pin not allowed --> that endpoint's constraint is dropped --> null path.

  So: worstSlack() = "search ran at all"; null-guard = "this specific endpoint has no path." Without the guard, PathExpanded expanded(path, ...) derefs null --> segfault. So both fixes needed.

File: src/rsz/src/Resizer.cc [hunk 2]
```diff
@@ -3245,6 +3245,81 @@ class SearchPredCombLogic : public sta::SearchPred1
...
+//         searchFrom(from_vertex) && searchThru(edge).
+// The mode-less SearchPred methods are used, which OR across all modes.
+// visit_seeds=false starts from the pred-filtered adjacency of each seed
+// (BfsIterator::enqueueAdjacentVertices seeding); visit_seeds=true visits
+// the seeds themselves first (BfsIterator::enqueue seeding).
+// visit(vertex) returns true if the search should expand through vertex.
+enum class DfsDirection
+{
+  fanin,
+  fanout
```
Comment (maliberty) [inline src/rsz/src/Resizer.cc]: https://google.github.io/styleguide/cppguide.html#Enumerator_Names
kFanin, kFanout

File: src/rsz/src/Resizer.cc [hunk 3]
```diff
@@ -3245,6 +3245,81 @@ class SearchPredCombLogic : public sta::SearchPred1
...
+};
+
+static void dfsSearch(sta::Graph* graph,
+                      sta::SearchPred& pred,
+                      DfsDirection dir,
```
Comment (maliberty) [inline src/rsz/src/Resizer.cc:3276]: const

**Review by maliberty — CHANGES_REQUESTED** [review]
minor points

**Review by maliberty — APPROVED** [review]
